peers/app_request_network.go

```diff
@@ -424,24 +433,82 @@ func (c *CanonicalValidators) GetValidator(nodeID ids.NodeID) (*warp.Validator,
 }
 
 // GetCanonicalValidators returns the validator information in canonical ordering for the given subnet
-// at the time of the call, as well as the total weight of the validators that this network is connected to
+// at the specified P-Chain height, as well as the total weight of the validators that this network is connected to
+// The caller determines the appropriate P-Chain height (ProposedHeight for current, specific height for epoched)
 func (n *appRequestNetwork) GetCanonicalValidators(
 	ctx context.Context,
 	subnetID ids.ID,
 	skipCache bool,
+	pchainHeight uint64,
 ) (*CanonicalValidators, error) {
-	// Get the subnet's current canonical validator set
-	fetchVdrsFunc := func(subnetID ids.ID) (avalancheWarp.CanonicalValidatorSet, error) {
-		startPChainAPICall := time.Now()
-		validatorSet, err := n.validatorClient.GetCurrentCanonicalValidatorSet(ctx, subnetID)
-		n.setPChainAPICallLatencyMS(float64(time.Since(startPChainAPICall).Milliseconds()))
-		return validatorSet, err
+	n.logger.Debug("Getting validator set at P-Chain height",
+		zap.Stringer("subnetID", subnetID),
+		zap.Uint64("pchainHeight", pchainHeight),
+		zap.Bool("isProposedHeight", pchainHeight == pchainapi.ProposedHeight),
+	)
+
+	var validatorSet avalancheWarp.CanonicalValidatorSet
+	var err error
+
+	if pchainHeight == pchainapi.ProposedHeight {
```

Comment:
We shouldn't ever use the `pchainapi.ProposedHeight` constant as a key in the cache, because the "proposed" height is constantly updating to represent new actual P-Chain heights.